### PR TITLE
BlockingRule: Clarify name of sql property

### PR DESCRIPTION
**Pull request by RobinL**
### Type of PR

- [ ] BUG
- [ ] FEAT
- [x] MAINT
- [ ] DOC

## Summary

This PR attempts to clarify the BlockingRule class.  It is a step towards enabling [ExplodingBlockingRules](https://github.com/moj-analytical-services/splink/pull/1692)

It will be the first in a set of three PRs which incrementally add clarifications.

### Clarify name of sql property
🤔 Previously `BlockingRule` had a `sql` property which was unused, and a property called `blocking_rule` which returned the sql.  The later was confusing: did it return a `BlockingRule` or a sql string?.  

**Solution**: There's now a single property called `blocking_rule_sql`

### Blocking rule __init__ argument allowed `blocking_rule` to be BlockingRule, dict or str

🤔 This made the code hard to read/unpredictable code because types were unclear

**Solution** It's now a string only

### Clarify name methos
🤔 Previously `BlockingRule` had a `sql` property `and_not_preceding_rules_sql(self)`.   Not very clear

**Solution**: I have clarified method names here to`exclude_pairs_generated_by_this_rule_sql` and `exclude_pairs_generated_by_all_preceding_rules_sql`.  This is partly because these implementations will differnt in #1692 

Finally I've also added type hinting to make refactoring a bit easier

### CODE REVIEW

File: splink/blocking.py [hunk 1]
```diff
@@ -40,15 +40,15 @@ def blocking_rule_to_obj(br):
 class BlockingRule:
     def __init__(
         self,
-        blocking_rule: BlockingRule | dict | str,
+        blocking_rule_sql: BlockingRule | dict | str,
```
Comment (RobinL) [inline splink/blocking.py:43]: I don't think we should allow this to take as an input any of those types - i think this should be string only

**Comment by ThomasHepworth** [comment]
A wider thought for another PR/Splink 4 - I think [`apply_salt`](https://github.com/moj-analytical-services/splink/blob/2bcad9be1988a3042123fecb47fc016054d409c7/splink/blocking.py#L268) should be handled from within the `BlockingRule` class, not outside.

It wouldn't be too hard to set the dialect for each BR in a central settings class/area when we go to unpack the settings dictionary (assuming we are planning to continue forward with that).

It's generally bad practice to have high 'coupling' (dependence between functions), which this is a perfect example of - changes to the salting methodology needs to be closely mirrored in the function linked above.

On top of that, it's quite hard to follow if you're new to the code base.

It would be better if we simply had to call `br.salted_blocking_rules` to get the output. Whether it came from a SparkLinker would then ideally be handled by the `BlockingRule` class (or a salting subclass).

**Review by ThomasHepworth — APPROVED** [review]
This looks good!

**Comment by ThomasHepworth** [comment]
My comment above is already being addressed by - https://github.com/moj-analytical-services/splink/pull/1701. Glad to see we're thinking the same thing.

File: splink/blocking.py [hunk 2]
```diff
@@ -40,15 +40,20 @@ def blocking_rule_to_obj(br):
 class BlockingRule:
     def __init__(
         self,
-        blocking_rule: BlockingRule | dict | str,
+        blocking_rule_sql: str,
         salting_partitions=1,
```
[the remark this answers is not in the document]
Comment (RobinL) [reply splink/blocking.py:44]: Good opint. I'll add it to the next PR, which changes this a bit anyway
